nuxeo-services/nuxeo-mail/src/main/java/org/nuxeo/mail/MailSessionBuilder.java

```diff
@@ -0,0 +1,227 @@
+
+                @Override
+                protected Session retrieveSession() {
+                    try {
+                        return FromJndi.this.retrieveSession();
+                    } catch (NuxeoException e) {
+                        log.debug("Lookup failed for javax.mail.Session with jndi name: {}, fallback on properties",
+                                jndiSessionName, e);
+                        return super.retrieveSession();
+                    }
+                }
+
+                @Override
+                public String toString() {
+                    return new ToStringBuilder(this).appendToString(FromJndi.this.toString())
+                                                    .appendToString(super.toString())
+                                                    .build();
+                }
+            };
+        }
+
+        @Override
+        public String toString() {
+            return new ToStringBuilder(this).append("jndiSessionName", jndiSessionName).toString();
+        }
+
+    }
+
+    public static class FromProperties extends AbstractFrom<FromProperties> {
+
+        protected Properties properties;
+
+        private FromProperties(Properties properties) {
+            this.properties = properties;
+            // remove the debug properties due to javax.mail.Session initialisation which prints to the console
+            this.debug = properties != null
+                    && Boolean.parseBoolean((String) properties.remove(CONFIGURATION_MAIL_DEBUG));
+        }
+
+        @Override
+        protected Session retrieveSession() {
+            log.debug("Lookup session from properties: {}", properties);
+            return Session.getInstance(properties, new MailAuthenticator(properties));
+        }
+
+        @Override
+        public String toString() {
+            ToStringBuilder builder = new ToStringBuilder(this);
+            Map<Object, Object> propertiesAsMap = properties.entrySet()
+                                                            .stream()
+                                                            .filter(s -> !s.getKey().toString().contains("password"))
+                                                            .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+            builder.append("properties", propertiesAsMap);
+            return builder.toString();
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    protected abstract static class AbstractFrom<F extends AbstractFrom> implements FromBuilder {
+
+        protected boolean debug;
+
+        /**
+         * Enables debug mode for built session.
+         *
+         * @implNote {@code mail.debug} in nuxeo.conf takes precedence if it's true.
+         * @return this builder
+         */
+        @SuppressWarnings("unchecked")
+        public F debug() {
+            this.debug = true;
+            return (F) this;
+        }
+
+        public Session build() {
```

Comment:
I got the buildAndConnect to explain there is a build before the connect. But if you have a public Session build() doesn't connect, why not making a `public void connect(Session session)` ? And then buildAndConnect() would just call both ?
I'm saying this, supposing you want to offer the possibility to build without connecting right away.

Reply:
That's not the same idea, `Session` doesn't have to be connected, Transport and Store have to.

I'm think about the purpose of `connect` method proposal, do you see one? I think we currently don't have such needs, `buildAndConnect` exists because we have a case in Nuxeo where we only need the Store.

We may have other methods like this in the future, like for instance getting a well configured `MimeMessage` ready to be sent with `Transport#send`. For now, I want to wait these usages to implement them correctly.

Reply:
OK I get it, Indeed I assumed Session need to be connected to work.